src/cudamatrix/cu-kernels.cu

```diff
@@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
   }
 }
 
+template<typename Real>
+__global__
+static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
+                                  Real* obfj_terms, MatrixDim objf_dim, 
+                                  Real* obfj_derivs, MatrixDim derivs_dim) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
+  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
+  int32_cuda scores_index = i + j * scores_dim.stride;
+  Real K = 1.0 / (scores_dim.rows - 2.0);
+  Real L = scores[scores_index];
+  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
```

Comment:
The fact that it was possible in thee forward computation generally means
it's possible i the backward computation.
You'll get S_deriv = X' diag(cvec_deriv) X, I think.


On Sat, Feb 13, 2016 at 5:39 PM, david-ryan-snyder <notifications@github.com
> wrote:

> In src/cudamatrix/cu-kernels.cu
> <https://github.com/danpovey/kaldi/pull/5#discussion_r52834702>:
>
> > @@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
> >    }
> >  }
> >
> > +template<typename Real>
> > +__global__
> > +static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
> > +                                  Real* obfj_terms, MatrixDim objf_dim,
> > +                                  Real* obfj_derivs, MatrixDim derivs_dim) {
> > +  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
> > +  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
> > +  int32_cuda scores_index = i + j * scores_dim.stride;
> > +  Real K = 1.0 / (scores_dim.rows - 2.0);
> > +  Real L = scores[scores_index];
> > +  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
>
> OK, I'll play with it some more to see if I can get it to work without a
> kernel and without an O(N^2) computation.
>
> In your procedure, it isn't obvious to me that you can get terms of the
> form S_deriv = x x' + y y' for all combinations of (x,y) pairs. That's
> where the O(N^2) comes from that I'm trying to avoid.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/5/files#r52834702>.
>